hrl/agent/dsc/classifier/sift_classifier.py

```diff
@@ -0,0 +1,322 @@
+        """
+        extract the SIFT features of a list of images
+        args:
+            images: a list of RGB/GrayScale images
+        return:
+            a list of SIFT features
+        """
+        keypoints = self.sift_detector.detect(images)
+        keypoints, descriptors = self.sift_detector.compute(images, keypoints)
+        return descriptors  # type: tuple
+    
+    def train_kmeans(self, sift_features):
+        """
+        train the kmeans classifier using the SIFT features
+        args:
+            sift_features: a list of SIFT features
+        """
+        # reshape the data
+        # each image has a different number of descriptors, we should gather 
+        # them together to train the clustering
+        sift_features=np.array(sift_features, dtype=object)
+        sift_features=np.concatenate(sift_features, axis=0).astype(np.float32)
+
+        # train the kmeans classifier
+        if self.kmeans_cluster is None:
+            if len(sift_features) < self.num_clusters:
+                # can't train kmeans with less example than clusters
+                # just duplicate the examples
+                num_duplicate = math.ceil(self.num_clusters / len(sift_features))
+                sift_features = np.tile(sift_features, (num_duplicate, 1))
+            self.kmeans_cluster = cluster.MiniBatchKMeans(n_clusters=self.num_clusters, random_state=0).fit(sift_features)
+        else:
+            self.kmeans_cluster.partial_fit(sift_features)
+    
+    def histogram_from_sift(self, sift_features):
+        """
+        transform the sift features by putting the sift features into self.num_clusters 
+        bins in a histogram. The counting result is a new feature space, which
+        is used directly by the SVM for classification
+        args:
+            sift_features: a list of SIFT features
+        
+        this code is optimized for speed, but it's doing essentially the following:
+        hist_features = []
+        for sift in sift_features:
+            # classification of all descriptors in the model
+            # sift.shape == (n_descriptors, 128)
+            predicted_cluster = self.kmeans_cluster.predict(sift)  # (n_descriptors,)
+            # calculates the histogram
+            # hist, bin_edges = np.histogram(predicted_cluster, bins=self.num_clusters)  # (num_clusters,)
+            hist = np.bincount(predicted_cluster, minlength=self.num_clusters)  # (num_clusters,)
+            # histogram is the feature vector
+            hist_features.append(hist)
+
+        hist_features = np.asarray(hist_features)
+        return hist_features
+        """
+        assert self.kmeans_cluster is not None, "kmeans classifier not trained"
+
+        n_descriptors_per_image = [len(sift) for sift in sift_features]
+        idx_num_descriptors = list(itertools.accumulate(n_descriptors_per_image))
+        sift_features_of_all_images = np.concatenate(sift_features, axis=0).astype(np.float32)
```

Comment:
Is this a legal operation given that `sift_features` can have variable number of columns per row?